src/tokenserver/db/models.rs

```diff
@@ -61,152 +70,294 @@ impl TokenserverDb {
+            .bind::<Text, _>(&params.email)
+            .bind::<Bigint, _>(timestamp)
+            .execute(&self.inner.conn)
+            .map(|_| ())
             .map_err(Into::into)
     }
 
-    #[cfg(test)]
-    fn post_service_sync(&self, service: params::PostService) -> DbResult<results::PostService> {
-        let query = r#"
-            INSERT INTO services (service, pattern)
-            VALUES (?, ?)
+    /// Mark the user with the given uid and service ID as being replaced.
+    fn replace_user_sync(&self, params: params::ReplaceUser) -> DbResult<results::ReplaceUser> {
+        const QUERY: &str = r#"
+            UPDATE users
+               SET replaced_at = ?
+             WHERE service = ?
+               AND uid = ?
         "#;
-        diesel::sql_query(query)
-            .bind::<Text, _>(&service.service)
-            .bind::<Text, _>(service.pattern)
-            .execute(&self.inner.conn)?;
 
-        let query = r#"
-            SELECT id
-              FROM services
+        diesel::sql_query(QUERY)
+            .bind::<Bigint, _>(params.replaced_at)
+            .bind::<Integer, _>(params.service_id)
+            .bind::<Bigint, _>(params.uid)
+            .execute(&self.inner.conn)
+            .map(|_| ())
+            .map_err(Into::into)
+    }
+
+    /// Update the user with the given email and service ID with the given `generation` and
+    /// `keys_changed_at`.
+    fn put_user_sync(&self, params: params::PutUser) -> DbResult<results::PutUser> {
+        // The `where` clause on this statement is designed as an extra layer of
+        // protection, to ensure that concurrent updates don't accidentally move
+        // timestamp fields backwards in time. The handling of `keys_changed_at`
+        // is additionally weird because we want to treat the default `NULL` value
+        // as zero. 
+        const QUERY: &str = r#"
+            UPDATE users
+               SET generation = ?,
+                   keys_changed_at = ?
              WHERE service = ?
-               AND pattern = ?
+               AND email = ?
+               AND generation <= ?
+               AND COALESCE(keys_changed_at, 0) <= COALESCE(?, 0)
+               AND replaced_at IS NULL
         "#;
-        diesel::sql_query(query)
-            .bind::<Text, _>(&service.service)
-            .bind::<Text, _>(&service.service)
-            .get_result::<results::PostService>(&self.inner.conn)
+
+        diesel::sql_query(QUERY)
+            .bind::<Bigint, _>(params.generation)
+            .bind::<Nullable<Bigint>, _>(params.keys_changed_at)
+            .bind::<Integer, _>(&params.service_id)
+            .bind::<Text, _>(&params.email)
+            .bind::<Bigint, _>(params.generation)
+            .bind::<Nullable<Bigint>, _>(params.keys_changed_at)
+            .execute(&self.inner.conn)
+            .map(|_| ())
             .map_err(Into::into)
     }
 
-    #[cfg(test)]
+    /// Create a new user.
```

Comment:
There's not really an analogous database method on the legacy Tokenserver. User "creation" is referred to as ["allocation"](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/assignment/sqlnode/sql.py#L368), and I'm not sure how I'm going to handle that yet. I may move some of the allocation logic to the handler

Reply:
The query I used is based off of the one [here](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/assignment/sqlnode/sql.py#L58)